File: go/sorting/insertion_sort.go
package sorting

type InsertionSort struct {
}

func (b InsertionSort) Sort(arr []int, sortType SortType) []int {
	length := len(arr)

	for i := 0; i < length; i++ {
		j := i
		for j >= 1 && Compare(arr[j-1], arr[j], sortType) {
			arr[j-1], arr[j] = arr[j], arr[j-1]
			j--
		}
	}

	return arr
}

File: dart/lib/sorting/insertion_sort.dart
import 'package:algorithm/sorting/sort.dart';

class InsertionSort extends Sort {
  @override
  List<int> sort(List<int> arr, {SortType type = SortType.asc}) {
    final length = arr.length;

    for (var i = 1; i < length; i++) {
      var j = i;
      while (j >= 1 && compare(arr[j - 1], arr[j], type)) {
        arr.swap(j - 1, j);
        j--;
      }
    }

    return arr;
  }
}
